sonic-thermalctld/scripts/thermalctld

```diff
@@ -0,0 +1,301 @@
+        presence = try_get(fan.get_presence, False)
+        speed = try_get(fan.get_speed)
+        speed_tolerance = try_get(fan.get_speed_tolerance)
+        speed_target = try_get(fan.get_target_speed)
+
+        if fan_name not in self.prensences:
+            self.prensences[fan_name] = True
+
+        if presence != self.prensences[fan_name]:
+            self._report_presense(fan_name, presence)
+            self.prensences[fan_name] = presence
+
+        fan_bitmap = FanUpdater.FAN_OK
+        if not presence:
+            fan_bitmap |= FanUpdater.FAN_NOT_PRESENCE
+        else:
+            fan_bitmap |= self._get_fan_bitmap(speed, speed_target, speed_tolerance)
+
+        self._set_fan_led(fan_bitmap)
+
+        fvs = swsscommon.FieldValuePairs(
+                [('presence', str(presence)),
+                ('model', str(try_get(fan.get_model))), 
+                ('serial', str(try_get(fan.get_serial))), 
+                ('status', str(try_get(fan.get_status, False))),
+                ('direction', str(try_get(fan.get_direction))),
+                ('speed', str(speed)),
+                ('speed_tolerance', str(speed_tolerance)),
+                ('speed_target', str(speed_target)),
+                ('led_status', str(try_get(fan.get_status_led))),
+                ('timestamp', datetime.now().strftime('%Y%m%d %H:%M:%S'))
+                ])
+        
+        self.table.set(fan_name, fvs)
+
+    def _report_presense(self, fan_name, presence):
+        if presence:
+            logger.log_warning('Fan removed warning cleared: {} was inserted.'.format(fan_name))
+        else:
+            logger.log_warning('Fan removed warning: {} was removed from the system, potential overheat hazard!'.format(fan_name))
+
+    def _get_fan_bitmap(self, speed, speed_target, speed_tolerance):
+        if speed != NOT_AVAILIABLE and speed_target != NOT_AVAILIABLE and speed_tolerance != NOT_AVAILIABLE:
+            return FanUpdater.FAN_UNEXPECT_SPEED if abs(speed - speed_target) > speed_tolerance else FanUpdater.FAN_OK
+        else:
+            return FanUpdater.FAN_OK
+
+    def _set_fan_led(self, fan, fan_bitmap):
+        if fan_bitmap == FanUpdater.FAN_OK:
+            fan.set_status_led(FanBase.STATUS_LED_COLOR_GREEN)
+        else:
+            fan.set_status_led(FanBase.STATUS_LED_COLOR_RED)
+
+        
+#
+# TemperUpdater ======================================================================
+#
+class TemperUpdater(object):
+    TEMPER_INFO_TABLE_NAME = 'TEMPERATURE_INFO'
+
+    def __init__(self, chassis):
+        self.chassis = chassis
+        self.temper_warnings = [(False, False)] * chassis.get_num_thermals()
+        state_db = daemon_base.db_connect(swsscommon.STATE_DB)
+        self.table = swsscommon.Table(state_db, TemperUpdater.TEMPER_INFO_TABLE_NAME)
+
+    def update(self):
+        logger.log_info("Start temperature updating")
+        for index, thermal in enumerate(self.chassis.get_all_thermals()):
```

Comment:
IMO, to use index returned by "enumerate" is a bit risky. It relies on an implicit assumption that there were a fixed mapping between "index" and object. But we can't say that for sure.
If not, the _refresh_temper_status will generate different name for an object in each iteration. For an exmaple:
1st iteration it returns {(0,"cpu_pack"), (1,"cpu_core0"),(2,"cpu_core1")} the names of "cpu_pack", "cpu_core0" and "cpu_core1" are "Thermal 0", "Thermal 1" and "Thermal 2" resprectively.
and 2nd iteration it returns {(0,"cpu_core0"),(1,"cpu_core1"),(2,"cpu_pack") then the name will change.

Reply:
This is a list, so the iteration order is stable. Since the "index" here is only for generate name for a thermal object whose "get_name" function not implemented, Stephen suggested to ignore those thermals lack of "get_name".

Reply:
I'd like to keep this. I think it would be better to "create" a name for any object who does not implement get_name than just ignore it. How do you think?

Reply:
yes, to give it a name is better than to ignore it. But I suggest using some fixed index, like the list index in the chassis object.